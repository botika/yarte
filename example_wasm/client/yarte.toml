# TODO: unique config file or composable
[main]
dir = "../templates"
debug = "code"

[debug]
short = false
