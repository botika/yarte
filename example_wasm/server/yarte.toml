[main]
dir = "../templates"

[debug]
short = false
