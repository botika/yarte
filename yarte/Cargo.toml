[package]
name = "yarte"
version = "0.5.1"
authors = ["Juan Aguilar Santillana <mhpoin@gmail.com>"]
description = "Type-safe, compiled Handlebars"
categories = ["template-engine"]
documentation = "https://docs.rs/yarte"
edition = "2018"
include = ["src/**/*", "Cargo.toml"]
keywords = ["markup", "template", "handlebars", "html"]
license = "MIT/Apache-2.0"
readme = "../README.md"
repository = "https://github.com/botika/yarte"
workspace = ".."

[badges]
travis-ci = { repository = "botika/yarte", branch = "master" }
maintenance = { status = "actively-developed" }

[features]
default = []
with-actix-web = ["actix-web", "futures", "yarte_derive/actix-web"]
full = ["with-actix-web"]

[dependencies]
yarte_config = { version = "0.5", path = "../yarte_config" }
yarte_derive = { version = "0.5", path = "../yarte_derive" }
yarte_helpers = { version = "0.5", path = "../yarte_helpers" }

actix-web = { version = "2.0", optional = true }
futures = { version = "0.3", optional = true }

yarte_wasm_app = { version = "0.0", path = "../yarte_wasm_app" }

[build-dependencies]
yarte_config = { version = "0.5", path = "../yarte_config" }
