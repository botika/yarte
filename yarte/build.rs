use yarte_helpers::recompile::when_changed;

fn main() {
    when_changed();
}
