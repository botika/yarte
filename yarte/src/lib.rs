//! Yarte stands for **Y**et **A**nother **R**ust **T**emplate **E**ngine,
//! is the fastest template engine. Uses a Handlebars-like syntax,
//! well known and intuitive for most developers. Yarte is an optimized, and easy-to-use
//! rust crate, can create logic around their templates using using conditionals, loops,
//! rust code and templates composition.
//!
//! Also Yarte incorporates feature `with-actix-web`, an implementation of `actix-web`'s
//! trait Responder for those using this framework.
//!
//! [Yarte book](https://yarte.netlify.com)
//!

#[cfg(not(target_arch = "wasm32"))]
use std::fmt::{self, Write};

pub use yarte_derive::Template;
#[cfg(not(target_arch = "wasm32"))]
pub use yarte_helpers::{helpers::Render, Error, Result};
#[cfg(target_arch = "wasm32")]
pub use yarte_wasm_app::App as Template;

pub mod recompile;

/// Template trait, will implement by derive like `Display` or `actix_web::Responder` (with feature)
///
/// ```rust
/// use yarte::Template;
///
/// #[derive(Template)]
/// #[template(src="Hello, {{ name }}!")]
/// struct HelloTemplate<'a> {
///     name: &'a str,
/// }
///
/// println!("{}", HelloTemplate { name: "world" })
/// ```
///
#[cfg(not(target_arch = "wasm32"))]
pub trait Template: fmt::Display {
    /// which will write this template
    fn call(&self) -> Result<String> {
        let mut buf = String::with_capacity(Self::size_hint());
        write!(buf, "{}", self).map(|_| buf)
    }

    /// https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types
    #[cfg(feature = "with-actix-web")]
    fn mime() -> &'static str
    where
        Self: Sized;

    /// Approximation of output size used in method `call`.
    /// Yarte implements an heuristic algorithm of allocation.
    fn size_hint() -> usize;
}

#[cfg(all(feature = "with-actix-web", not(target_arch = "wasm32")))]
pub mod aw {
    pub use actix_web::{
        error::ErrorInternalServerError, Error, HttpRequest, HttpResponse, Responder,
    };
    pub use futures::future::{err, ok, Ready};
}

#[cfg(all(feature = "wasm", not(target_arch = "wasm32")))]
pub mod serde_json {
    pub use serde_json::to_string;
}
