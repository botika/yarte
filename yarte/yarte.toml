# root dir of templates
[main]
dir = "templates"

# Alias for partials. In call, change the start of partial path with one of this, if exist.
[partials]
alias = "example/deep/more/deep"
doc = "example/deep/more/doc"
card = "example/deep/more/card"
