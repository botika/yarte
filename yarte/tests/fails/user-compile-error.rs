use yarte::Template;

#[derive(Template)]
#[template(src = "{{#if true}} {{$ \"foo\" }} {{/if }}")]
struct Test{
    foo: usize
}

fn main() {}
