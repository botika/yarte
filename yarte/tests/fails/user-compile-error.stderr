error: proc-macro derive panicked
 --> $DIR/user-compile-error.rs:3:10
  |
3 | #[derive(Template)]
  |          ^^^^^^^^
  |
  = help: message: error
           --> $DIR/Test.hbs:1:14
            |
          1 | {{#if true}} {{$ "foo" }} {{/if }}
            |              ^^^^^^^^^^^^ Compile error: foo
            |
